### PR TITLE
Add Hungarian algorithm for the linear assignment problem.

### PR DESCRIPTION
**Pull request by carlosgmartin**
#954

### CODE REVIEW

**Comment by fabianp** [comment]
thanks for starting this @carlosgmartin !

At a high level this looks good to me, but it would be important to have an example of this method in the gallery (https://optax.readthedocs.io/en/latest/gallery.html) to showcase its usage

**Review by vroulet — COMMENTED** [review]
Thanks for integrating this
Here are a few comments

File: optax/assignment/_hungarian_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+# ==============================================================================
+"""The Hungarian algorithm for the linear assignment problem."""
+
+from functools import partial
+
+import jax
+from jax import lax, numpy as jnp
+
+
+def hungarian_algorithm(costs, /):
```
[the remark this answers is not in the document]
Comment (carlosgmartin) [reply optax/assignment/_hungarian_algorithm.py:23]: There's only 1 argument and I was undecided between calling it `costs`, `cost_matrix`, `matrix`, `cost_array`, `array`, etc. so I figured we can just keep the API name-independent for now.

If you prefer a specific argument name, let me know.

File: optax/assignment/_hungarian_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+  )
+
+  if transpose:
+    i = col4row.argsort()
+    return col4row[i], i
+  else:
+    return jnp.arange(costs.shape[0]), col4row
+
+
+def _find_short_augpath_while_body_inner_for(it, val):
```
[the remark this answers is not in the document]
Comment (carlosgmartin) [reply optax/assignment/_hungarian_algorithm.py:75]: Added a reference to the paper with the mathematical expressions: https://ieeexplore.ieee.org/document/7738348

**Comment by carlosgmartin** [comment]
@fabianp @vroulet Done.

File: optax/assignment/_hungarian_algorithm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vroulet) [reply optax/assignment/_hungarian_algorithm.py:23]: I see. Interesting solution, I'll keep it in mind in the future.
I think for now it's ok to use `cost_matrix` (Wikipedia uses such term so we'll consider it standard).
I'd prefer a name (potentially not optimal) than an uncommon signature.

File: optax/assignment/_hungarian_algorithm_test.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import optax
+
+
+class HungarianAlgorithmTest(parameterized.TestCase):
+
+  @parameterized.product(
+    n=[0, 1, 2, 4, 8, 16],
+    m=[0, 1, 2, 4, 8, 16],
+  )
+  def test_hungarian_algorithm(self, n, m):
```
Comment (vroulet) [inline optax/assignment/_hungarian_algorithm_test.py:30]: It would be great to see whether this algorithm can support usual transformations (like jitting or vmap).

Side note beyond this PR:
It could be great to have this building block differentiable. To be exact: there is no way the current outputs can be differentiable (they are ints). But the resulting cost could be differentiable. The advantage is that people may then parameterize the cost matrix, run the algorithm to get the optimal cost and then differentiate through the algorithm to further refine their cost matrix.
This would not be super hard actually to let your function return the optimal cost so that a user could easily make it differentiable. 
Anyway, it's not very important, so you can dismiss this comment

File: optax/assignment/_hungarian_algorithm.py [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+    v,
+    shortest_path_costs,
+    path,
+    lowest,
+    row4col,
+    index,
+  )
+
+
+def _find_short_augpath_while_body_tail_alt(val):
```
Comment (vroulet) [inline optax/assignment/_hungarian_algorithm.py:181]: All names of functions are very well chosen, which makes the code nicely readable. An exception is "alt" here: is it for alternative? Either change this word to something readable or add a comment below

**Review by vroulet — COMMENTED** [review]
Thanks again! A few minor remaining comments

File: optax/assignment/_hungarian_algorithm.py (hunk 4 above)
Comment (carlosgmartin) [reply optax/assignment/_hungarian_algorithm.py:181]: Ah, that's supposed to be just `_find_short_augpath_while_body_tail`. (I was testing an alternative implementation.)

File: optax/assignment/_hungarian_algorithm_test.py (hunk 3 above)
Comment (carlosgmartin) [reply optax/assignment/_hungarian_algorithm_test.py:31]: Currently, the algorithm has 2 occurrences of [lax.while_loop](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.while_loop.html), which is not reverse-mode differentiable.

In theory, these could be replaced with a [bounded while_loop](https://github.com/jax-ml/jax/issues/2469). I don't think that's been ported [from jaxopt](https://github.com/google/jaxopt/blob/main/jaxopt/_src/loop.py) yet, has it?

On the other hand, it seems to me that, for [almost all](https://en.wikipedia.org/wiki/Almost_all) inputs and infinitesimal perturbations, the (integer) solution indices remain unchanged, and thus differentiating the solution cost is equivalent to simply differentiating `cost_matrix[i_solution, j_solution]`.

**Comment by carlosgmartin** [comment]
@vroulet Updated.

File: optax/assignment/_hungarian_algorithm_test.py (hunk 3 above)
Comment (vroulet) [reply optax/assignment/_hungarian_algorithm_test.py:31]: > has 2 occurrences of lax.while_loop

Ok, I should have been more careful, this is indeed an impediment for an efficient differentiation.
Anyway, we may think about that later, the current implementation should already be useful for many use cases.
And thanks for checking the interoperability with jit and vmap!

**Review by vroulet — APPROVED** [review]
Great work, looks good to me

**Comment by vroulet** [comment]
Yes, that's clearly worth it :)
Ideally we would like your effort to be adopted so such comments are good.
Also, I was wondering, your implementation is neat, but aren't there other implementations potentially more tpu friendly? (not very important, but one may mention it if it's the case? I'm not an expert on that, packages in optimal transport may have some clever tricks).

**Comment by carlosgmartin** [comment]
@vroulet Right, there are faster specialized algorithms for approximating the Wasserstein distance (see [OTT-JAX](https://github.com/ott-jax/ott)).

Perhaps I can append the example above to the [LAP gallery entry](https://optax.readthedocs.io/en/latest/_collections/examples/linear_assignment_problem.html)?

**Comment by JTT94** [comment]
How does this compare to `hungarian_tpu_matcher`? from 
https://github.com/google-research/scenic/blob/main/scenic/model_lib/matchers/hungarian_jax.py

**Comment by carlosgmartin** [comment]
@vroulet Submitted a PR: #1140.
